src/libstd/sys/common/io.rs

```diff
@@ -0,0 +1,133 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use prelude::v1::*;
+use io;
+use io::ErrorKind;
+use io::Read;
+use slice::from_raw_parts_mut;
+
+// Provides read_to_end functionality over an uninitialized buffer.
+// This function is unsafe because it calls the underlying
+// read function with a slice into uninitialized memory. The default
+// implementation of read_to_end for readers will zero out new memory in
+// the buf before passing it to read, but avoiding this zero can often
+// lead to a fairly significant performance win.
+//
+// Implementations using this method have to adhere to two guarantees:
+//  *  The implementation of read never reads the buffer provided.
+//  *  The implementation of read correctly reports how many bytes were written.
+pub unsafe fn read_to_end_uninitialized<R: Read + ?Sized>(mut r: &mut R, buf: &mut Vec<u8>)
+                                                          -> io::Result<usize> {
+
+    let start_len = buf.len();
+    buf.reserve(16);
```

Comment:
Actually, now that I think about it, I think we can remove this call to `reserve` because of the `reserve(1)` below. The same logic should apply that if the vector isn't full we've got space to read into and if it is full then a new reservation will double its size.

Also, can you add a comment to the loop along the lines of:

> Always try to read into the empty space of the vector (from the length to the capacity). If the vector ever fills up then we reserve an extra byte which should trigger the normal reallocation routines for the vector which will likely double the size.
> 
> This function is similar to the `read_to_end` function in std::io, but the logic about reservations and slicing is different enough that this is duplicated here.